subiquity/common/api/client.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2020 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import inspect
+import json
+
+import aiohttp
+
+from subiquitycore import contextlib38
+
+from subiquity.common.serialize import Serializer
+from .defs import Payload
+
+
+def _wrap(make_request, path, meth, serializer):
+    sig = inspect.signature(meth)
+    meth_params = sig.parameters
+    payload_arg = None
+    for name, param in meth_params.items():
+        if getattr(param.annotation, '__origin__', None) is Payload:
+            payload_arg = name
+            payload_ann = param.annotation.__args__[0]
+    r_ann = sig.return_annotation
+
+    async def impl(*args, **kw):
+        args = sig.bind(*args, **kw)
+        params = {
+            k: json.dumps(serializer.serialize(meth_params[k].annotation, v))
+            for (k, v) in args.arguments.items() if k != payload_arg
+            }
+        if payload_arg in args.arguments:
+            v = args.arguments[payload_arg]
+            data = serializer.serialize(payload_ann, v)
```

Comment:
Is there any sort of lurking bug if payload_ann doesn't get set in the _wrap call? Or do we feel confident that Payload will always be one of the parameter's annotations

Reply:
No, this is actually fine because if there is no Payload argument, then payload_arg will be None and so cannot be in arg.arguments, which is a string-keyed dictview. But it is a bit unclear, and I was already a bit unhappy with it, so I'll try a rewrite.